mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/events/WorkerMetricsCollector.java

```diff
@@ -0,0 +1,213 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.mantisrx.master.events;
+
+import com.netflix.spectator.api.Tag;
+import io.mantisrx.common.metrics.Metrics;
+import io.mantisrx.common.metrics.MetricsRegistry;
+import io.mantisrx.common.metrics.Timer;
+import io.mantisrx.common.metrics.spectator.MetricGroupId;
+import io.mantisrx.master.events.LifecycleEventsProto.JobStatusEvent;
+import io.mantisrx.master.events.LifecycleEventsProto.WorkerListChangedEvent;
+import io.mantisrx.master.events.LifecycleEventsProto.WorkerStatusEvent;
+import io.mantisrx.master.jobcluster.job.worker.IMantisWorkerMetadata;
+import io.mantisrx.master.jobcluster.job.worker.WorkerState;
+import io.mantisrx.server.core.domain.WorkerId;
+import io.mantisrx.server.master.domain.JobId;
+import io.mantisrx.shaded.com.google.common.util.concurrent.AbstractScheduledService;
+import io.mantisrx.shaded.org.apache.curator.shaded.com.google.common.base.Preconditions;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import lombok.RequiredArgsConstructor;
+import lombok.Value;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * The goal of this service is to keep emitting metrics around how long it took for the worker to be
+ * scheduled, to be prepared and how long it was running for.
+ */
+@RequiredArgsConstructor
+@Slf4j
+public class WorkerMetricsCollector extends AbstractScheduledService implements
+    WorkerEventSubscriber {
+
+    public static final String SCHEDULING_DURATION = "schedulingDuration";
+    public static final String PREPARATION_DURATION = "preparationDuration";
+    public static final String RUNNING_DURATION = "runningDuration";
+    private final ConcurrentMap<JobId, Map<WorkerId, IMantisWorkerMetadata>> jobWorkers =
+        new ConcurrentHashMap<>();
+    private final ConcurrentMap<String, WorkerMetrics> clusterWorkersMetrics =
+        new ConcurrentHashMap<>();
+    private final Set<CleanupJobEvent> jobsToBeCleaned = new HashSet<>();
+    private final Duration cleanupInterval;
+
+    private final Duration epochDuration;
+    private final Clock clock;
+
+    @Override
+    protected void runOneIteration() throws Exception {
+        Instant expiry = clock.instant().minus(cleanupInterval);
+        Set<CleanupJobEvent> toBeRemoved = new HashSet<>();
+
+        synchronized (jobsToBeCleaned) {
```

Comment:
wondering if you could use a ConcurrentHashMap for the set as well
and also using an iterator on jobToBeCleaned and call remove on the iterator to remove.w

Reply:
good point. let me do that.

Reply:
done.